File: week05/19/P3.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def __init__(self):
        self.head = None

    def insert(self, data):
        new_node = ListNode(data)
        new_node.next = self.head
        self.head = new_node

    def print_list(self):
        curr = self.head
        while (curr is not None):
            print(curr.val, end=" -> ")
            curr = curr.next
        print("nullptr")

    def middleNode(self):
        fast = self.head
            
        while (fast is not None and fast.next is not None):
            fast = fast.next.next
            self.head = self.head.next

if __name__ == "__main__":
    head = Solution()
    head.insert(6)
    head.insert(5)
    head.insert(4)
    head.insert(3)
    head.insert(2)
    head.insert(1)
    head.print_list()  # 1 -> 2 -> 3 -> 4 -> 5 -> 6 -> nullptr

    head.middleNode()
    head.print_list()  # 4 -> 5 -> 6 -> nullptr

File: week05/19/P3.cpp
#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

ListNode* insert_head(ListNode *head, int val) {
    ListNode *new_node = new ListNode(val, head);
    return new_node;
};

void printList(ListNode* head) {
    ListNode* current_node = head;
    while (current_node != nullptr) {
        cout << current_node->val << " -> ";
        current_node = current_node->next;
    }
    cout << "nullptr" << endl;
}

class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        ListNode *fast = head;
        ListNode *slow = head;
        
        while (fast != nullptr && fast->next != nullptr) {
            fast = fast->next->next;
            slow = slow->next;
        }

        return slow;
    }
};

int main() {
    ListNode *head = nullptr;
    head = insert_head(head, 6);
    head = insert_head(head, 5);
    head = insert_head(head, 4);
    head = insert_head(head, 3);
    head = insert_head(head, 2);
    head = insert_head(head, 1);
    printList(head);  // 1 -> 2 -> 3 -> 4 -> 5 -> 6 -> nullptr
    
    Solution result;
    head = result.middleNode(head);
    printList(head);  // 4 -> 5 -> 6 -> nullptr

    return 0;
}